### PR TITLE
Cancel edit branch

### PR DESCRIPTION
**Pull request by jkilfoyle**
Added cancellation, prescription redeeming and prescription updating to blockchain and backend.

Included an index checker, for valid indeces on the blockchain, currently not hooked into any backend functions, but available for future extension.

Truffle tests for redeeming, cancellation, and updating.
Blockchain script for cancellation, and update included. 

Backend functionality for cancellation tested, updating still needs to be tested through the backend.
I did not see a backend route for redeeming, but it could be that the backend was intended to implement redeeming through update, in which case the redeem function is redundant.

This is a large update, so review carefully before merge, thanks.

### CODE REVIEW

**Comment by jacobkrantz** [comment]
Not ready for merging yet. @mdulin2 let me run some tests before you review this.

**Comment by mdulin2** [comment]
@jacobkrantz  how does this look, Jacob?

File: WebPhapp/PharmaChain/contracts/PrescriptionData.sol [hunk 1]
```diff
@@ -5,48 +5,103 @@ import "./PrescriptionBase.sol";
...
-        uint64[16] memory fulfillmentDates,
-        uint64 dateWritten,  
-        uint16 daysValid,
-        uint8 refillsLeft,
-        bool isCancelled, 
-        uint64 cancelDate) public returns (uint) {
-        Prescription memory p = Prescription(patientID,prescriberID,dispenserID,drugID,drugQuantity,
-        fulfillmentDates,dateWritten, daysValid, refillsLeft, isCancelled, cancelDate);
+
+    function addPrescription(uint256 patientID, uint128 prescriberID, uint128 dispenserID, uint64 drugID, string memory drugQuantity,
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/contracts/PrescriptionData.sol]: I personally disagree with the styling change here. I think it is much easier to read in the previous format.

File: WebPhapp/PharmaChain/contracts/PrescriptionData.sol [hunk 2]
```diff
@@ -5,48 +5,103 @@ import "./PrescriptionBase.sol";
...
+        if (p.refillsLeft == 0) {
+            //cannot refill empty prescription
+            return 2;
+        }
+        p.refillsLeft = p.refillsLeft - 1;
+        for(uint64 i = 0; i < 16; i++) {
+            if(p.fulfillmentDates[i] == 0) {
+                p.fulfillmentDates[i] = date;
+                break;
+                //NOTE we need to discuss what happens if initial refillsLeft > 16
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/contracts/PrescriptionData.sol]: Good call...

File: WebPhapp/PharmaChain/test/TestPrescriptionData.sol [hunk 3]
```diff
@@ -13,7 +13,7 @@ contract TestPrescriptionData is PrescriptionBase{
     
     Patient p = Patient(DeployedAddresses.Patient());
 
-    uint64[16] fu; // cannot assign arrays as a constant. So, this var must be here.
+    uint64[16] fu = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]; // cannot assign arrays as a constant. So, this var must be here.
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/test/TestPrescriptionData.sol]: Good change.

File: WebPhapp/PharmaChain/test/TestPrescriptionData.sol [hunk 4]
```diff
@@ -51,6 +51,59 @@ contract TestPrescriptionData is PrescriptionBase{
...
         Assert.equal(string(drugQuantity), string(data.drugQuantity), "Drug quantity error....");
     }
+
+	//Test ability to update a prescription
+    function testUpdate() public{
+		uint index = p.addPrescription(data.patientID, data.prescriberID, data.dispenserID, data.drugID,
+        data.drugQuantity, data.fulfillmentDates, data.dateWritten, data.daysValid, data.refillsLeft,
+        data.isCancelled, data.cancelDate);
+	p.updatePrescription(index, 3, data.drugQuantity, data.daysValid, data.refillsLeft);
+
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/test/TestPrescriptionData.sol]: Why the awk formatting here?

File: WebPhapp/PharmaChain/truffle.js [hunk 5]
```diff
@@ -20,8 +20,7 @@ module.exports = {
 			host: "10.50.0.2",
 			port: 22000,
 			network_id: "*",
-			gasPrice: 0,
-			gas: 999999909
+			gasPrice: 0
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/truffle.js]: Purpose of this?

File: WebPhapp/PharmaChain/update_prescription.js [hunk 6]
```diff
@@ -2,13 +2,20 @@ let fs = require("fs");
 let Web3 = require("web3");
 let net = require("net");
 
-/*  This function updates an existing prescription on the drugChain
-    User input: prescription arguments
-    Argument list: drugChainIndex, dispenserID, drugQuantity, daysValid, isCancelled, cancelDate
-    Example usage: node update_prescription.js 0 1 2 4 true 8    
+/*  
+This function updates an existing prescription on the drugChain
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/update_prescription.js]: Good commenting.

File: WebPhapp/PharmaChain/update_prescription.js [hunk 7]
```diff
@@ -2,13 +2,20 @@ let fs = require("fs");
...
+    dispenserID (int)
+    drugQuantity (string)
+    daysValid (int)
+    refillsLeft (int)
+Example: 
+    >>> sudo node update_prescription.js 0 1 "1 mg" 4 8
 */
 
-async function update( drugChainIndex, dispenserID,drugQuantity, fulfillmentDate, daysValid, isCancelled, cancelDate){
+async function update( drugChainIndex, dispenserID, drugQuantity, daysValid, refillsLeft){
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/update_prescription.js]: Why was fulfillment dates removed?

File: WebPhapp/PharmaChain/contracts/PrescriptionData.sol [hunk 8]
```diff
@@ -5,48 +5,103 @@ import "./PrescriptionBase.sol";
...
+    function isValidIndex(uint chainIndex) public view returns (uint) {
+        if(0 <= chainIndex && chainIndex < getDrugChainLength()) {
+            return 0;
+        }
+        //invalid chain index
+        return 1;
+    }
 
-        return 0;
+    function updatePrescription(uint chainIndex, uint128 dispenserID, string drugQuantity,
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/contracts/PrescriptionData.sol]: Should we validate which fields can be changed here?

File: WebPhapp/PharmaChain/update_prescription.js [hunk 9]
```diff
@@ -18,7 +25,7 @@ async function update( drugChainIndex, dispenserID,drugQuantity, fulfillmentDate
     account = account[0];
 
     // Sets up deployment requirements
-    let source = fs.readFileSync('/home/osboxes/PharmaChain/build/contracts/Patient.json'); 
+    let source = fs.readFileSync('./build/contracts/Patient.json');
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/update_prescription.js]: Good change!

File: WebPhapp/PharmaChain/update_prescription.js [hunk 10]
```diff
@@ -28,8 +35,7 @@ async function update( drugChainIndex, dispenserID,drugQuantity, fulfillmentDate
 
     // Set up prescription data to be sent.
     Patient.options.address = fs.readFileSync("./patient_contract_address.txt").toString('ascii');
-    let transaction = await Patient.methods.updatePrescription(drugChainIndex, dispenserID,drugQuantity, 
-                                                        fulfillmentDate, daysValid, isCancelled, cancelDate);
+    let transaction = await Patient.methods.updatePrescription(drugChainIndex, dispenserID, drugQuantity, daysValid, refillsLeft);
```
Comment (mdulin2) [inline WebPhapp/PharmaChain/update_prescription.js]: @jacobkrantz  are you okay with awaits being used here? They are so much easier to use than promises, in my opinion.

File: WebPhapp/backend/block_helper.js [hunk 11]
```diff
@@ -169,15 +169,133 @@ module.exports = {
...
+	decrementing refills left, and adding a new date to fulfillmentDates
+    Args:
+        chainIndex (int)
+        date (int)
+    Returns:
+        Transaction object
+    Example:
+        redeem(0, 123456543)
+    */
+    redeem: async function(chainIndex, date) {
```
Comment (mdulin2) [inline WebPhapp/backend/block_helper.js]: At what point are the dates converted back to integers from strings on the frontend? @jacobkrantz

File: WebPhapp/backend/block_helper.js [hunk 12]
```diff
@@ -169,15 +169,133 @@ module.exports = {
...
+        update(0, 2, '300MG', 16, 1)
+    */
+    update: async function(chainIndex, dispenserID, drugQuantity, daysValid, refillsLeft) {
+        var blockchain = await connectToChain();
+        var error;
+        var block;
+        try {
+            let transaction = await blockchain.patient.methods.updatePrescription(
+                chainIndex,
+                dispenserID,
```
Comment (mdulin2) [inline WebPhapp/backend/block_helper.js]: DispensorID is editable?

**Comment by mdulin2** [comment]
Overall, great job @jkilfoyle and @jacobkrantz! I just had some thoughts.

**Comment by mdulin2** [comment]
Is there a reason this is still open? It's been 6 days... I would love to see this merged...

File: WebPhapp/backend/block_helper.js (hunk 11 above)
Comment (jacobkrantz) [reply WebPhapp/backend/block_helper.js]: `convertDatesToString()` in `server.js` should always do this.

**Comment by jacobkrantz** [comment]
@jkilfoyle check this out. We need to have merged for meeting tomorrow. I created the redeem route in `server.js` and tested all the new functions. They should be good to go, but you should test them yourself first. You can run curl commands with the examples in routes `/api/v1/prescriptions/cancel/:prescriptionID`,`/api/v1/prescriptions/edit`, and `/api/v1/dispensers/redeem/:prescriptionID`. I did not add unit tests because they only affect the blockchain. Maybe you could think about how we can better tests the routes?

**Comment by jkilfoyle** [comment]
@jacobkrantz I loaded up the changes, but unfortunately only the cancel curl request worked succesfully for me. With edit and update I encountered the following backend error:
/api/v1/prescriptions/edit: error: TypeError: values.5.map is not a function
/api/v1/dispensers/redeem: error: TypeError: values.5.map is not a function

Hopefully this is just a problem on my end, which we can clear up. But for now I can't complete this pull request until this is ironed out.

**Comment by jacobkrantz** [comment]
@jkilfoyle I tried on my machine with a fresh start. Works as expected.
Steps:
Remove your quorum folder, reboot your machine, run `testRunner.sh`, and set blockchain to true in `connections.js`. Then all three routes should be testable. I found running them in update->redeem->cancel order to be the easiest (this way I never had to add a new prescription).

**Pull request merged**